### PR TITLE
Upgrade to joi v15, make joi normal dependency

### LINKED ISSUE #21: Update to Joi v15+ (the latest)
tsdv-joi requires an old version of Joi and its type definitions, let's update to the latest.

This will introduce breaking changes. For example, Joi removed all "date" functionality from the core library, to remove the dependency on moment, thereby minimising the library footprint and making it easier to package for the browser. For tsdv-joi, this means we'd need a separate project to provide decorators for [joi-date](https://github.com/hapijs/joi-date).

### LINKED ISSUE #22: Make Joi a regular dependency (not a peer dependency)
Joi was originally required as a peer dependency, for easier integration with existing codebases that already used Joi. I don't think that provides much value now. Let's just package Joi (and @types/joi) as a regular dependency, and export Joi from tsdv-joi, so that consuming apps don't also need to have their own dependency on Joi (and @types/joi) for constructing schemas for e.g. the `@Items()` decorator.

Relates to #21.

### PR DESCRIPTION
**Pull request by codeandcats**
### Summary
- Upgrades joi to v15 (`@hapi/joi`)
  - Was actually pretty painless
  - A couple tests needed updating as joi fixed a couple issues and the fixes broke the tests
  - Didn't run into any issues with dates. Seems date validation is still supported by v15.
- Made joi a regular dependency (not a peer dependency)

Closes #21 and closes #22.

### CODE REVIEW

File: src/validation.ts [hunk 1]
```diff
@@ -96,7 +96,10 @@ export class Validator {
             joi.validate(target, classSchema, options) :
             joi.validate(target, classSchema);
 
-        return result as ValidationResult<TInstance>;
+        return {
+            ...result,
+            value: result.value as TInstance,
```
Comment (codeandcats) [inline src/validation.ts:101]: P.S. Here we're pretending that `Partial<TInstance>` is actually a `TInstance`. Same below.

File: test/unit/core.test.ts [hunk 2]
```diff
@@ -65,14 +66,20 @@ describe('getJoiSchema', () => {
             fluffinessIndex: jf.joi.number().min(1).max(5),
         });
 
-        expect(JSON.stringify(getJoiSchema(Cat, jf.joi))).toEqual(JSON.stringify(expectedSchema));
+        expect(stringify(getJoiSchema(Cat, jf.joi))).toEqual(stringify(expectedSchema));
```
Comment (codeandcats) [inline test/unit/core.test.ts:69]: New joi object has circular references so had to use a stringify library that doesn't explode on circular refs.

File: test/unit/decorators/any.test.ts [hunk 3]
```diff
@@ -294,15 +294,15 @@ describe('any', () => {
...
+            @jf.string().meta({ value: 'some metadata' })
             name?: string;
         }
 
         it('should associate a label with the property', () => {
             assertClassSchemaEquals({
                 Class: User,
                 expectedSchemaMap: {
-                    name: jf.joi.string().meta('some metadata'),
+                    name: jf.joi.string().meta({ value: 'some metadata' }),
```
Comment (codeandcats) [inline test/unit/decorators/any.test.ts:305]: It didn't like me passing in a `string` here, wanted an `object` instead.

File: test/unit/decorators/string.test.ts [hunk 4]
```diff
@@ -63,9 +63,9 @@ describe('string', () => {
             [
                 { emailAddress: 'monkey@see.com' },
                 { emailAddress: 'howdy+there@pardner.co.kr' },
-                { emailAddress: 'monkey@do' },
             ],
             [
+                { emailAddress: 'monkey@do' },
```
Comment (codeandcats) [inline test/unit/decorators/string.test.ts:68]: Joi now requires email addresses to have a top level domain.

File: test/unit/decorators/string.test.ts [hunk 5]
```diff
@@ -239,12 +239,11 @@ describe('string', () => {
                 },
                 [
                     { ipAddress: '127.0.0.1/24' },
+                    { ipAddress: '2001:db8:abcd:8000::/50' },
```
Comment (codeandcats) [inline test/unit/decorators/string.test.ts:242]: Joi now supports...this.

File: test/unit/validation.test.ts [hunk 6]
```diff
@@ -28,7 +28,7 @@ describe('ValidationResult', () => {
                     {
                         message: "'email' is not a valid email",
                         type: 'email',
-                        path: 'emailAddress',
+                        path: ['emailAddress'],
```
Comment (codeandcats) [inline test/unit/validation.test.ts:31]: `path` is now an array.

File: test/unit/decorators/string.test.ts (hunk 5 above)
Comment (laurence-myers) [reply test/unit/decorators/string.test.ts:242]: It's an IPv6 subnet, I think

**Comment by laurence-myers** [comment]
Regarding dates, looks like joi only split out the `format()` (and `utc()`) validation, which was the only thing that required `moment`. We don't have a decorator for `format()`, so no problems I think. 🙂 

https://github.com/hapijs/joi-date/blob/master/API.md

**Pull request merged**